src/three-components/SmoothControls.ts

```diff
@@ -409,7 +412,12 @@ export class SmoothControls extends EventDispatcher {
     // the spherical->position derivation:
     if (!this[$previousPosition].equals(this.camera.position)) {
       this[$previousPosition].copy(this.camera.position);
-      this.dispatchEvent({type: 'change'});
+
+      const source = this[$isUserChange] ? 'user-interaction' : '';
+
+      this.dispatchEvent({type: 'change', source});
+
+      this[$isUserChange] = false;
```

Comment:
I think this needs to be part of the implementation of `setOrbit`. Otherwise, `$isUserChange` will be set to false after one incremental movement of the camera.

Reply:
I sort of think this is ok. Because this is when the actual physical rotation happened. If we dispatched in `setOrbit` it would be when the rotation was initiated.

One thing we could do here is something like:
```
this[$isUserChange] = this[$pointerIsDown] || false;
```

Reply:
I may be mistaken, but I believe that as implemented this will only dispatch a single event per user interaction. Is that the intention?

Reply:
Currently yes. But it seemed like you didn't think was the best

> I think this needs to be part of the implementation of setOrbit. Otherwise, $isUserChange will be set to false after one incremental movement of the camera.

I do think if we use the event `change` with `source` then it should still fire at the time when the camera orbit has been resolved. Or it has "changed".

Reply:
As implemented, this change is sufficient to support the feature under consideration, because you only need to know the first time there was some change due to user interaction. But...

If we are decorating the `'change'` event, I would like this to be more coherent with how `SmoothControls` works in general. If the camera moves on account of user interaction, it moves to a target destination over multiple steps (except for very close targets, or cases where a frame blows its budget and we move multiple steps at once). Either every step towards a target destination is provoked by user interaction, or none of them are.

Two coherent approaches that I can imagine are:

 1. Every `'change'` event that is dispatched during the camera's approach to the target destination that was set by user interaction is marked with `'user-interaction'` as its `source`
 2. A "start" and "end" event of some kind are dispatched to indicate when changes on account of user interaction begin and end

I like the first one better. It is simpler to handle as a consumer of the `SmoothControls` API (there is only one event that we need to listen to). We are already dispatching the event anyway, it's just a matter of decorating it with some extra context.

---

Stepping outside of the box for a second, we could try reframing the goal: maybe all we care about is when the _target destination_ is changed. We currently don't have an event for this, but perhaps we need one?

Note that this might amount to sending twice the events in a lot of cases. The target destination can change on every `mousemove` or `touchmove` in practice.

Thinking about it this way might be a better way for us to close the gap in our ideas here. I'm interested to hear your thoughts on this.

Reply:
I like 1) better also.

Maybe we could change `this[$userAdjustOrbit]` to be something like this:
```
    private[$userAdjustOrbit](deltaTheta: number, deltaPhi: number, deltaRadius: number): boolean {
    const handled = this.adjustOrbit(deltaTheta, deltaPhi, deltaRadius);

    if (this[$userTargetSpherical] === null) {
      this[$userTargetSpherical] = new Spherical();
    }

    this[$userTargetSpherical].copy(this[$targetSpherical]);

    return handled;
  }
```

This way we know where the user is wanting to go. If `setOrbit` is called then the `$targetSpherical` won't match `$userTargetSpherical` and the event won't be dispatched. At this point `this[$userTargetSpherical]` could be made `null`.

I think the reason why `this[$userTargetSpherical]` should be `null` initially is that we can't initialize it to `0, 0, 0` because it may inadvertently dispatch `source: user-interaction` erroneously.

Does this make sense?

`Spherical` does not have a `equals` method but it should be pretty easy to compare these two spherical coordinates in `update`.

Reply:
@mikkoh that makes sense, but I think it's unnecessary. If `setOrbit` sets the flag to `false`, and then `userSetOrbit` sets the flag to `true` immediately after, you would have the same amount of information without creating an additional spherical instance (and comparing all of its values later), right?

Reply:
@cdata I explored this a bit here https://github.com/mikkoh/model-viewer/pull/2

Reply:
That looks fine at a glance, but I'm still not convinced that comparing sphericals is necessary.

We have two cases where the target destination changes:

 1. The API `setOrbit` is invoked imperatively by some embedding code (sometimes indirectly via `adjustOrbit`)
 2. The user interacts with `<model-viewer>`, causing `$userAdjustOrbit` to be called (indirectly invoking `adjustOrbit` etc)

Since we only have these two cases, we can _always_ set the flag to `false` when `setOrbit` is called. And, we can _always_ set the flag to `true` when `$userAdjustOrbit` is called.

If we did this, we wouldn't need to compare sphericals because the boolean flag contains all the information we need.

Did I get that right? Maybe I missed something important...

Reply:
This was actually my initial idea too but I wasn't sure how future scalable it would be.

There is one issue though when `setOrbit` is called with the exact same values coming from `$userAdjustOrbit`. In that case I'm assuming it should be a user interaction still. Thoughts?

Reply:
I think I agree with you. Let's construct a practical scenario and you can tell me what you think is best:

 1. The user interacts and orbits to spherical target `foo`, causing `auto-rotate` to temporarily stop
 2. Then, the user waits for `1` second
 3. Then, a timer fires that causes JavaScript to imperatively set the orbit to spherical target `foo`

Assuming that the total `auto-rotate` delay would normally be `3` seconds, how many seconds should pass after step 3 before `auto-rotate` starts again?

If we say that it's still a user interaction, only `2` more seconds should pass before `auto-rotate` starts again. Otherwise, `3` more seconds should pass.

Reply:
I think it should start counting down the 3 seconds when it has arrived at `foo`. Because if the user interaction is large enough and we've talked about 3 seconds being variable (via attributes) there could be a chance that there is no user delay.

Since you seem to agree with keeping track of a user target I'll rebase https://github.com/mikkoh/model-viewer/pull/2

Unless there are objections.

Reply:
Noo, sorry, I thought we had moved passed that and agreed on a boolean. My previous response was to this comment you made:

> There is one issue though when setOrbit is called with the exact same values coming from $userAdjustOrbit. In that case I'm assuming it should be a user interaction still.

Reply:
But to be able to set that boolean in `setObit` you need to know the user target.

With:
> There is one issue though when setOrbit is called with the exact same values coming from $userAdjustOrbit. In that case I'm assuming it should be a user interaction still.

If you simply just set `isUserChange` to `false` in `setOrbit` then the application might inadvertently cancel the user interaction. An example of this:

1. User touch drags the model
2. `$userAdjustOrbit` is called which calls `setOrbit`
3. `setOrbit` is executed and `isUserChange` is set to `false`
4. `$userAdjustOrbit` continues executing and sets `isUserChange` to `true`
5. `update` ticks dispatching 'change' with 'source: user-interaction'
6. `setOrbit` is called because of HTML attributes changing but it happens that the new target orbit is set to the same target orbit as set by `$userAdjustOrbit`
7. `setOrbit` changes `isUserChange` to `false` even though the requested target by `setOrbit` is the same as set by the previous `$userAdjustOrbit`
8. Update ticks dispatching 'change' with 'source: none' although it's animating to the same user targeted position.

In the above PR the flow is like this:
1. User touch drags the model
2. `$userAdjustOrbit` is called which calls `setOrbit`
3. `setOrbit` is executed and `isUserChange` is set evaluated to either `true` or `false` depending on if `userTargetSpherical` and `targetSpherical` match. In this case they did not match.
4. `$userAdjustOrbit` continues executing and sets `isUserChange` to `true` and also `userTargetSpherical` is made to match `targetSpherical`
5. `update` ticks dispatching 'change' with 'source: user-interaction'
6. `setOrbit` is called because of HTML attributes changing but it happens that the new target orbit is set to the same target orbit as set by `$userAdjustOrbit`
7. `setOrbit` evaluates `isUserChange` to either `true` or `false` depending on if `userTargetSpherical` and `targetSpherical` match. In this case they did match.
8. `update` ticks dispatching 'change' with 'source: user-interaction'

All this to say I think we need a `boolean` and `Spherical`. 

Another reason why we need `boolean` and `Spherical` is because on init the user target spherical will be initialized as `0, 0, 0` however this not actually a user target. It's just the initial values. The `boolean` enforces that the user target spherical has been truly set. Another way to handle this would be to set the user target spherical to `null`. But in this case `setOrbit` would set the `userTargetSpherical` to `null` and immediately after `$userAdjustOrbit` would allocate memory again for the `userTargetSpherical`. Another thing is that when `source` is evaluated in `update` it's much simpler to compare the Boolean `isUserChange` vs comparing target and user target Sphericals.

Sorry this is a bit complicated 😬

Reply:
Thanks for thinking about this so carefully @mikkoh . I do see what you are saying about an imperative call to `setOrbit` potentially overriding the boolean (and thus overriding the user).

For the sake of simplicity, I think we can just say that if `setOrbit` is invoked imperatively, it implies that the content author wants a very specific result (for example, they picked a preset point of interest to orbit to), and it's okay for that to be considered an override of the user's choice.

Reply:
Ok so what you're saying is that we should just set `isUserChange` to `false` on `setOrbit`.

A few days out I'm ok with this. I'll update this PR.